library/core/src/slice/mod.rs

```diff
@@ -76,6 +76,24 @@ pub use index::SliceIndex;
 #[unstable(feature = "slice_check_range", issue = "76393")]
 pub use index::check_range;
 
+/// Calculates the direction and split point of a one-sided range.
+///
+/// Helper for `take` and `take_mut` which returns a boolean
+/// indicating whether the front of the split is being taken
+/// (as opposed to the back), as well as a number indicating the
+/// index at which to split. Returns `None` if the split index would
+/// overflow `usize`.
+#[inline]
+fn take_split_point(range: impl OneSidedRange<usize>) -> Option<(bool, usize)> {
```

Comment:
Wouldn't it be better to make `take_split_point` a method of `OneSideRange` and implement it for ranges? Not only it is type safe, it would also somewhat speed up debug builds since compiler wouldn't have to emit code for match and panic.

Reply:
I believe type safety wise there's no difference since this function also only takes `OneSidedRange`s, but otherwise that sounds reasonable. It would mean that `take_split_point` becomes part of `OneSidedRange`'s API, but that's probably fine since it's unstable anyway?

Reply:
Hmm, `take_split_point` in its current form can't really be implemented for any `OneSidedRange<T>`, because of the `checked_add` calls. Any idea what kind of API we could add to `OneSidedRange` that offers the same functionality without being restricted to integers?

Reply:
I don't think that this restriction can be a concern in practice since at he moment you can't indes slice with ranges of something other than `usize`s anyway.

Reply:
Sure, but it'd be a bit weird to implement `OneSidedRange` only for `usize` ranges?

Reply:
If it bothers you that much, you can just name it `OneSidedUsizeRange` ¯\\\_(ツ)\_/¯

Reply:
I think it's best to leave this as a private function for now.

If this were to be a public API (even just unstable), it'd probably need a better return type, instead of just using `bool` to indicate the direction. As a private function, that's fine though.